File: src/org/elixir_lang/reference/Module.java
package org.elixir_lang.reference;

import com.intellij.codeInsight.lookup.LookupElement;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiPolyVariantReference;
import com.intellij.psi.PsiReferenceBase;
import com.intellij.psi.ResolveResult;
import com.intellij.psi.impl.source.resolve.ResolveCache;
import org.elixir_lang.psi.QualifiableAlias;
import org.elixir_lang.psi.scope.module.Variants;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

public class Module extends PsiReferenceBase<QualifiableAlias> implements PsiPolyVariantReference {
    /*
     * Fields
     */

    @NotNull
    public final PsiElement maxScope;

    /*
     * Constructors
     */

    public Module(@NotNull QualifiableAlias qualifiableAlias, @NotNull PsiElement maxScope) {
        super(qualifiableAlias, TextRange.create(0, qualifiableAlias.getTextLength()));
        this.maxScope = maxScope;
    }

    /*
     *
     * Instance Methods
     *
     */

    /*
     * Public Instance Methods
     */

    @NotNull
    @Override
    public ResolveResult[] multiResolve(boolean incompleteCode) {
        return ResolveCache
                .getInstance(this.myElement.getProject())
                .resolveWithCaching(
                        this,
                        org.elixir_lang.reference.resolver.Module.INSTANCE,
                        false,
                        incompleteCode
                );
    }

    @Nullable
    @Override
    public PsiElement resolve() {
        ResolveResult[] resolveResults = multiResolve(false);
        return resolveResults.length == 1 ? resolveResults[0].getElement() : null;
    }

    @NotNull
    @Override
    public Object[] getVariants() {
        List<LookupElement> lookupElementList = Variants.lookupElementList(myElement);

        return lookupElementList.toArray(new Object[lookupElementList.size()]);
    }
}

File: src/org/elixir_lang/reference/Module.kt
package org.elixir_lang.reference

import com.intellij.codeInsight.lookup.LookupElement
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiPolyVariantReference
import com.intellij.psi.PsiReferenceBase
import com.intellij.psi.ResolveResult
import com.intellij.psi.impl.source.resolve.ResolveCache
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.util.IncorrectOperationException
import org.elixir_lang.psi.ElementFactory
import org.elixir_lang.psi.ElixirAlias
import org.elixir_lang.psi.QualifiableAlias
import org.elixir_lang.psi.QualifiedAlias
import org.elixir_lang.psi.scope.module.Variants

class Module(qualifiableAlias: QualifiableAlias) :
        PsiReferenceBase<QualifiableAlias>(qualifiableAlias, textRange(qualifiableAlias)),
        PsiPolyVariantReference {
    @Throws(IncorrectOperationException::class)
    override fun handleElementRename(newElementName: String): PsiElement {
        val replacementAlias = PsiTreeUtil.findChildOfType(
            ElementFactory.createFile(myElement.project, newElementName),
            QualifiableAlias::class.java
        ) ?: throw IncorrectOperationException("Unable to parse module alias name: $newElementName")

        return myElement.replace(replacementAlias)
    }

    override fun getVariants(): Array<LookupElement> = Variants.lookupElements(myElement).toTypedArray()

    override fun isReferenceTo(element: PsiElement): Boolean {
        val manager = this.element.manager

        return multiResolve(false).any {
            it.isValidResult && manager.areElementsEquivalent(it.element, element)
        }
    }

    override fun multiResolve(incompleteCode: Boolean): Array<ResolveResult> {
        ApplicationManager.getApplication().assertReadAccessAllowed()
        return ResolveCache
            .getInstance(this.myElement.project)
            .resolveWithCaching(
                this,
                org.elixir_lang.reference.resolver.Module,
                false,
                incompleteCode
            )
    }

    override fun resolve(): PsiElement? = multiResolve(false).singleOrNull()?.element
}

private fun textRange(qualifiableAlias: QualifiableAlias): TextRange = when (qualifiableAlias) {
    is ElixirAlias -> TextRange.create(0, qualifiableAlias.textLength)
    is QualifiedAlias -> qualifiableAlias.getAlias().textRangeInParent
    else -> TODO()
}
